qupath-core/src/main/java/qupath/lib/images/ImageData.java

```diff
@@ -134,27 +139,56 @@ public String toString() {
 	
 	/**
 	 * Create a new ImageData with a specified object hierarchy and type.
-	 * @param server
-	 * @param hierarchy 
-	 * @param type
+	 * @param serverBuilder builder to use if the server is to be loaded lazily; may be null to se server instead
+	 * @param server server to use directly; may be null to use builder instead
+	 * @param hierarchy an object hierarchy, or null to create a new one
+	 * @param type the image type, or null to default to ImageType.UNSET
+	 * @throws IllegalArgumentException if neither a server nor a server builder is provided
 	 */
-	public ImageData(ImageServer<T> server, PathObjectHierarchy hierarchy, ImageType type) {
-		pcs = new PropertyChangeSupport(this);
+	private ImageData(ImageServerBuilder.ServerBuilder<T> serverBuilder, ImageServer<T> server, PathObjectHierarchy hierarchy, ImageType type)
+			throws IllegalArgumentException {
+		if (server == null && serverBuilder == null)
+			throw new IllegalArgumentException("Cannot create ImageData without a server or server builder");
+		this.pcs = new PropertyChangeSupport(this);
+		this.serverBuilder = serverBuilder;
 		this.server = server;
 		this.hierarchy = hierarchy == null ? new PathObjectHierarchy() : hierarchy;
-		this.serverPath = server == null ? null : server.getPath(); // TODO: Deal with sub image servers
 		initializeStainMap();
 		if (type == null)
 			type = ImageType.UNSET;
 		setImageType(type);
-		
+
 		// Add listeners for changes
 		this.hierarchy.addListener(this);
 		workflow.addWorkflowListener(this);
-		
+
 		// Discard any changes during construction
 		changes = false;
 	}
+
+	/**
+	 * Create a new ImageData with a lazily-loaded server, hierarchy and type.
+	 * The server builder provides the ImageServer required to access pixels and metadata on demand.
```

Comment:
_"[...] required to access pixels ~and metadata~ on demand."_

Reply:
Why remove `and metadata`? I was thinking of `ImageServer.getMetadata()`, which provides essential info.